lib/metrics.py

```diff
@@ -348,3 +346,77 @@ def _get_db_conn(self):
                 psycopg2.extensions.ISOLATION_LEVEL_AUTOCOMMIT
             )
         return self.db
+
+
+class PaidAppsMetricsEmitterThread(BaseMetricsEmitterThread):
+    @property
+    def _select_stats_to_emit(self):
+        selected_stats = []
+        if buildpackutil.i_am_primary_instance():
+            selected_stats = [
+                self._inject_database_stats,
+                self._inject_storage_stats,
+                self._inject_health,
+            ]
+        selected_stats.append(self._inject_m2ee_stats)
+        return selected_stats
+
+    def _gather_metrics(self):
+        stats = {}
+        try:
+            for inject_method in self._select_stats_to_emit:
+                stats = inject_method(stats)
+        except psycopg2.OperationalError as up:
+            logger.exception("METRICS: error while gathering metrics")
+            stats = {
+                "health": {
+                    "health": 0,
+                    "diagnosis": "Database error: {}".format(str(up)),
+                }
+            }
+        except Exception:
+            logger.exception("METRICS: error while gathering metrics")
+            stats = {
+                "health": {
+                    "health": 4,
+                    "diagnosis": "Unable to retrieve metrics",
+                }
+            }
+        finally:
+            return stats
+
+
+class FreeAppsMetricsEmitterThread(BaseMetricsEmitterThread):
+    def _get_munin_stats(self):
+        m2ee_stats, _ = munin.get_stats_from_runtime(
+            self.m2ee.client, self.m2ee.config
+        )
+        return m2ee_stats
+
+    def _inject_user_session_metrics(self, stats):
+        session_metrics = {}
+        try:
+            m2ee_stats = self._get_munin_stats()
+            if "sessions" in m2ee_stats.keys():
+                m2ee_stats["sessions"]["user_sessions"] = {}
+                session_metrics = m2ee_stats["sessions"]
+        except Exception:
```

Comment:
As mentioned above, we should be very careful when introducing extra buildpack logging, as it is exposed to customers.

In this case, I think that we do not need to log anything here. If runtime metrics aren't emitted, it's not worth putting into a customer's logs. (Even for a free app).

Reply:
Aren't we interested in reasons for failure? Since all logs go through CF Loggregator we can easily get insight on the root reason(s) for runtime metrics not being fetched/emitted. :thinking:

Reply:
No, we are not interested.

This is because the reasons are usually performance of customer apps, which is outside of our control.

They will do CPU/memory intensive stuff as part of their application, which will cause the runtime not to respond. This is normal behaviour. It's is not our responsibility (as operators of the Mendix cloud), but the responsibility of the customer (as the operator of their application).

We do not want to log anything regarding a _symptom_ of the problem (e.g. a timeout talking to the runtime), but instead want to expose stability problems via our usual mechanisms (i.e. alerts)

Reply:
This also has a heavy cost for us too, as we end up investigating these issues (in response to support tickets) and concluding that it is a performance problem for which the customer is responsible.

Reply:
As I mention above, the ideal situation here is something like Sentry where we can privately aggregate such errors in the buildpack and investigate if we deem necessary. But we do not have such a facility now, and won't have it any time soon.

Reply:
As discussed in person I have lowered the log level for those messages. In case internal log aggregator fo customer apps are in place we can still utilize those log messages, and they still stay 'hidden' from Free Apps cluster end users.